vpr/src/analytical_place/read_atom_netlist.cpp

```diff
@@ -0,0 +1,175 @@
+    // Each net has the exact same name as in the atom netlist
+    for (AtomBlockId atom_blk_id : atom_netlist.blocks()) {
+        // Get the molecule of this block
+        t_pack_molecule* mol = prepacker.get_atom_molecule(atom_blk_id);
+        // Create the AP block (if not already done)
+        const std::string& first_blk_name = atom_netlist.block_name(mol->atom_block_ids[0]);
+        APBlockId ap_blk_id = ap_netlist.create_block(first_blk_name, mol);
+        // Add the ports and pins of this block to the supernode
+        for (AtomPortId atom_port_id : atom_netlist.block_ports(atom_blk_id)) {
+            BitIndex port_width = atom_netlist.port_width(atom_port_id);
+            PortType port_type = atom_netlist.port_type(atom_port_id);
+            const std::string& port_name = atom_netlist.port_name(atom_port_id);
+            const std::string& block_name = atom_netlist.block_name(atom_blk_id);
+            // The port name needs to be made unique for the supernode (two
+            // joined blocks may have the same port name)
+            std::string ap_port_name = block_name + "_" + port_name;
+            APPortId ap_port_id = ap_netlist.create_port(ap_blk_id, ap_port_name, port_width, port_type);
+            for (AtomPinId atom_pin_id : atom_netlist.port_pins(atom_port_id)) {
+                BitIndex port_bit = atom_netlist.pin_port_bit(atom_pin_id);
+                PinType pin_type = atom_netlist.pin_type(atom_pin_id);
+                bool pin_is_const = atom_netlist.pin_is_constant(atom_pin_id);
+                AtomNetId pin_atom_net_id = atom_netlist.pin_net(atom_pin_id);
+                const std::string& pin_atom_net_name = atom_netlist.net_name(pin_atom_net_id);
+                APNetId pin_ap_net_id = ap_netlist.create_net(pin_atom_net_name);
+                ap_netlist.create_pin(ap_port_id, port_bit, pin_ap_net_id, pin_type, pin_is_const);
+            }
+        }
+    }
+
+    // Fix the block locations given by the VPR constraints
+    for (APBlockId ap_blk_id : ap_netlist.blocks()) {
+        const t_pack_molecule* mol = ap_netlist.block_molecule(ap_blk_id);
+        for (AtomBlockId mol_atom_blk_id : mol->atom_block_ids) {
+            PartitionId part_id = constraints.get_atom_partition(mol_atom_blk_id);
+            if (!part_id.is_valid())
+                continue;
+            // We should not fix a block twice. This would imply that a molecule
+            // contains two fixed blocks. This would only make sense if the blocks
+            // were fixed to the same location. I am not sure if that is even
+            // possible.
+            VTR_ASSERT(ap_netlist.block_mobility(ap_blk_id) == APBlockMobility::MOVEABLE);
+            // Get the partition region.
+            const PartitionRegion& partition_pr = constraints.get_partition_pr(part_id);
+            VTR_ASSERT(partition_pr.get_regions().size() == 1 && "Each partition should contain only one region!");
```

Comment:
Partitions can contain more than one region (we take the union as the legal locations).  If the analytic placer can't handle that you can assert for now (should handle or turn into a proper error later) but the assert message should mention this is an AP limitation.